File: frontend/apps/manage/src/lib/stages.ts
import type { EffortBandRef, TicketStage } from "../api";

/**
 * The client half of a state machine that has NO status column.
 *
 * The server derives a ticket's stage from the RIGHTMOST STAMPED of five
 * nullable timestamps, floored at `intake`. A NULL earlier stamp means "never
 * separately recorded", not "impossible" — forward skips are legal and audited.
 * So the order below is not decoration: it is the only place the console knows
 * what "later" means, and it must agree member-for-member with
 * `TicketStage`'s DECLARATION ORDER on the server, which is what the conditional
 * writes' `AND <every column after the target> IS NULL` predicate is spelled
 * from.
 *
 * In `lib/` so this section and F42's capacity picker share it with no import
 * cycle — `lib/booking.tsx`'s header gives the general reason.
 */
export const STAGE_ORDER: readonly TicketStage[] = [
  "intake",
  "in_progress",
  "qc",
  "ready",
  "delivered",
];

/**
 * The stage word for every member of the union.
 *
 * ⚠ `Record<TicketStage, string>` IS THE POINT, and it is why this is a record
 * rather than a function with a `default`. A sixth stage added to the union
 * without a key here is a COMPILE error; a switch with a fallback is a wrong
 * label that ships silently — the defect `lib/roles.ts` exists to record.
 *
 * The type catches a missing MEMBER; `i18n.test.ts` resolves every value so a
 * member pointing at a key absent from `he.ts` is caught too. The two halves
 * catch different bugs: a column rendering the raw slug `in_progress` and a
 * column rendering the wrong Hebrew word are not the same failure.
 */
export const STAGE_LABEL_KEY: Record<TicketStage, string> = {
  intake: "atelier.stage.intake",
  in_progress: "atelier.stage.inProgress",
  qc: "atelier.stage.qc",
  ready: "atelier.stage.ready",
  delivered: "atelier.stage.delivered",
};

/**
 * The skip `Select`'s options: stages STRICTLY later than the card's current
 * one, in board order, and the empty array at `delivered`.
 *
 * Strictly later, because a backwards skip is a 409 the server refuses and
 * re-entering the current stage is a 200 no-op that would announce a move that
 * did not happen. Offering either would be the console inviting a refusal.
 */
export function laterStages(current: TicketStage): TicketStage[] {
  return STAGE_ORDER.slice(STAGE_ORDER.indexOf(current) + 1);
}

/**
 * The stored minutes read back as a band word — a reverse lookup, because what
 * persists is `effort_minutes` and never the label.
 *
 * The fallback is the visible consequence of that rule: a boutique that
 * re-tuned «חצי יום» from 240 to 300 leaves every ticket estimated under the old
 * mapping matching no live band, and the card says «240 דק׳» rather than
 * silently re-valuing the garment. F41 ships no editor for the mapping (F42
 * owns it), so the fallback is reachable the day after F42 ships and not before.
 *
 * First match wins. The server bounds each band to 1..1440 and forbids no
 * duplicate, so an owner may flatten her two longest bands onto one number.
 */
export function bandLabel(
  minutes: number,
  bands: readonly EffortBandRef[],
  t: (key: string, options?: Record<string, unknown>) => string,
): string {
  const match = bands.find((band) => band.minutes === minutes);
  return match === undefined
    ? t("atelier.effortMinutes", { minutes })
    : t(`atelier.band.${BAND_KEY_SUFFIX[match.band]}`);
}

// The wire's snake_case band keys spelled as the camelCase tail of their i18n
// key. A `Record<EffortBand, string>` for the reason STAGE_LABEL_KEY is one: a
// sixth band without a suffix here is a compile error, not «atelier.band.xl»
// rendering its own key into a card.
const BAND_KEY_SUFFIX: Record<EffortBandRef["band"], string> = {
  thirty_min: "thirtyMin",
  one_hour: "oneHour",
  two_hours: "twoHours",
  half_day: "halfDay",
  full_day: "fullDay",
};

File: backend/app/atelier/stages.py
"""The pure core of the atelier board: the stage derivation, the predicate
builder every conditional write is spelled from, and the tenant band resolver.

No I/O, no session, no fakes — every function here is a total function of its
arguments, which is what lets the four-outcome discriminators in the service and
the guarded UPDATEs in the repository be tested without a database.

It lives here rather than in the repository or the service because BOTH need it:
the repository builds its predicates from `later_columns`, the service
discriminates its outcomes with `stage_of`, and putting either in the other would
create a backwards import edge.
"""

from typing import Any

from app.models.alteration_ticket import AlterationTicket
from app.models.constants import EffortBand, TicketStage
from app.models.staff_user import StaffUser

# One column per stage, in declaration order. The mapping is separate from the
# enum because the enum is the WIRE contract and these are column names — F44
# reads the columns directly and must not have to know the wire words.
STAGE_COLUMNS: dict[TicketStage, str] = {
    TicketStage.INTAKE: "intake_at",
    TicketStage.IN_PROGRESS: "in_progress_at",
    TicketStage.QC: "qc_at",
    TicketStage.READY: "ready_at",
    TicketStage.DELIVERED: "delivered_at",
}


def stage_of(row: AlterationTicket) -> TicketStage:
    """The RIGHTMOST stamped column, in declaration order.

    Total by construction: the fallback is INTAKE, so a row with no stamp at all
    — which the writer cannot produce, because the INSERT stamps intake_at — is
    read as intake rather than crashing a poll every five seconds. `StaffCard`'s
    `card_status` is the same shape one file over: a total function of nullable
    columns whose output set is pinned by a set-equality test.

    ⚠ RIGHTMOST, NEVER THE FIRST NULL. A seamstress who takes a hem from intake
    straight to ready in one sitting leaves in_progress_at and qc_at NULL forever
    and the ticket is at `ready`. An implementation that walks forward looking for
    the first NULL answers `in_progress` on that row — and answers it in the
    direction that lets a stale board stamp a stage the garment has already left.
    An earlier NULL beside a later stamp means "that stage was never separately
    recorded", not that it did not happen.
    """
    current = TicketStage.INTAKE
    for stage in TicketStage:
        if getattr(row, STAGE_COLUMNS[stage]) is not None:
            current = stage
    return current


def later_columns(target: TicketStage) -> tuple[str, ...]:
    """Every stamp column AFTER `target` in declaration order.

    This is D3's entire concurrency mechanism, and it does two jobs in one
    predicate the DATABASE evaluates: it refuses a backwards stamp (a stale board
    tapping `qc` on a ticket already at `ready`) and it refuses the loser of a
    genuine race — neither of which a pre-read can do, because another
    transaction can invalidate a pre-read between the SELECT and the UPDATE.

    `delivered` answers the empty tuple: nothing comes after it, so the last
    advance is an ordinary conditional write rather than a special case.

    The target's OWN column is deliberately absent — it is a separate clause,
    `IS NULL` for an advance and `IS NOT NULL` for an undo, and folding it in
    here would make the undo's predicate self-contradictory.
    """
    stages = list(TicketStage)
    return tuple(STAGE_COLUMNS[stage] for stage in stages[stages.index(target) + 1 :])


# The platform's five. A tenant tunes them per band; nobody has to configure
# anything for the board to work, and the E9 brief's reason for tunability at all
# ("'half-day' is not 240 minutes in a boutique whose shifts are six hours") is
# about F42's capacity arithmetic, which F41 does not perform.
DEFAULT_EFFORT_BANDS: dict[EffortBand, int] = {
    EffortBand.THIRTY_MIN: 30,
    EffortBand.ONE_HOUR: 60,
    EffortBand.TWO_HOURS: 120,
    EffortBand.HALF_DAY: 240,
    EffortBand.FULL_DAY: 480,
}

# The DB CHECK's own ceiling, restated here because this resolver is the ONLY
# thing between a hand-edited JSONB blob and that constraint. Without it a stored
# 5000 reaches the INSERT and answers a 500 on intake instead of a ticket.
MAX_BAND_MINUTES = 1440


def _positive_int(value: Any, default: int) -> int:
    """A positive int within the CHECK's bound, or the platform default.

    `bool` is excluded explicitly because it is an `int` subclass in Python and
    `True` would otherwise resolve to a one-minute band. Floats are refused
    rather than rounded: rounding picks a value nobody chose.
    """
    if isinstance(value, bool) or not isinstance(value, int):
        return default
    if value <= 0 or value > MAX_BAND_MINUTES:
        return default
    return value


def effort_bands(settings: dict[str, Any]) -> dict[EffortBand, int]:
    """The tenant's five bands, resolved from `tenants.settings`.

    ⚠ THIS READ COSTS NO STATEMENT. `TenantContext.settings` is already bound on
    the request by the tenancy middleware, so the router passes the dict straight
    in. Reading it back through `TenantsRepository` would open a FOURTH session,
    pool checkout and BEGIN/COMMIT on the hottest read in the feature, every five
    seconds per device — that repository is constructed with a session_factory and
    opens its own session inside every method, so it cannot join the atelier's
    tenant_session.

    Resolution is PER BAND with a platform default, so a partial mapping is legal
    and one junk value costs that band its tuning and nothing else. Iteration is
    over the ENUM and never over the stored dict, so a renamed key or a typo in a
    hand-edited blob cannot put a sixth band on the wire.

    **Every tenant always has exactly five bands.** That is what lets the intake
    form render with no empty-state branch and what lets `effort_minutes NOT NULL`
    hold. A brand-new boutique has no `atelier` key at all and that is the normal
    case, not an error — no shipped writer can reach the key, because
    `merge_settings` takes only `profile=` and `toggles=`. F42 owns the editor.
    """
    atelier = settings.get("atelier")
    stored = atelier.get("effort_bands") if isinstance(atelier, dict) else None
    if not isinstance(stored, dict):
        stored = {}
    return {
        band: _positive_int(stored.get(band.value), DEFAULT_EFFORT_BANDS[band])
        for band in EffortBand
    }


# 0022's CHECK ceiling, restated once so the column bound and the settings bound
# cannot drift. `boutique/validation.py` imports it from HERE and not from
# `atelier/validation.py`: this module imports only `app.models`, so the edge
# stays acyclic, while `atelier/validation.py` would drag `app.booking.validation`
# and `app.catalog.validation` in behind it (D5).
MAX_WEEKLY_CAPACITY_HOURS = 168


def default_capacity_hours(settings: dict[str, Any]) -> int | None:
    """The tenant's house default, or None. THERE IS NO PLATFORM DEFAULT.

    The asymmetry with `effort_bands`' five is deliberate (D2) and it is two
    arguments: `effort_minutes` is NOT NULL so band resolution must always answer
    a number, while `weekly_capacity_hours` is nullable and "unknown" is
    representable; and a wrong band is bounded inside five coarse presets while a
    wrong capacity is a DENOMINATOR that renders every one of her bars at a
    fraction of the truth, in the colour that says everything is fine.

    ⚠ IT DOES NOT SHARE `_positive_int`. That helper's range is `1..1440` and it
    refuses `0`; this one's is `0..168` and `0` is a value it must keep — "the
    boutique is stood down this week". One helper with two range parameters would
    be a parameterised abstraction over two call sites whose only interesting
    property is precisely the boundary it would have to parameterise.
    """
    atelier = settings.get("atelier")
    stored = atelier.get("default_weekly_capacity_hours") if isinstance(atelier, dict) else None
    if isinstance(stored, bool) or not isinstance(stored, int):
        return None
    return stored if 0 <= stored <= MAX_WEEKLY_CAPACITY_HOURS else None


def resolve_capacity(row: StaffUser, tenant_default: int | None) -> tuple[int | None, bool]:
    """Returns (resolved_hours, capacity_is_default).

    ⚠ `is not None`, NEVER truthiness. `0` is a DELIBERATE value — "she is not
    available this week" (D1) — and `or` would hand her the boutique's default,
    render her bar at a fraction of the truth in the non-overload colour, print
    «ברירת מחדל של הבוטיק» on a number she personally set, and sort her FIRST in
    the assign Select. The shipped `settings.get("atelier") or {}` idiom is
    exactly the habit that produces the bug here, which is why this is code and
    not a sentence.

    `capacity_is_default` is False when the answer is None: there is nothing to
    have defaulted to, and the row renders «לא הוגדרה קיבולת» with no bar.
    """
    if row.weekly_capacity_hours is not None:
        return row.weekly_capacity_hours, False
    return tenant_default, tenant_default is not None
